crates/ruff_python_parser/src/lib.rs

```diff
@@ -388,48 +388,94 @@ impl Tokens {
+    /// 5) `18..27` would panic because both the start and end offset is within a token
     ///
     /// ## Note
     ///
     /// The returned slice can contain the [`TokenKind::Unknown`] token if there was a lexical
     /// error encountered within the given range.
+    ///
+    /// # Panics
+    ///
+    /// If either the start or end offset of the given range is within a token range.
     pub fn tokens_in_range(&self, range: TextRange) -> &[Token] {
-        let Ok(start) = self.binary_search_by_key(&range.start(), Ranged::start) else {
-            return &[];
-        };
-        let Ok(end) = self[start..].binary_search_by_key(&range.end(), Ranged::end) else {
-            return &[];
-        };
-        &self[start..=start + end]
+        let tokens_after_start = self.after(range.start());
+        let end = tokens_after_start.partition_point(|token| token.end() <= range.end());
+        let result = &tokens_after_start[..end];
+
+        // If the end offset isn't the end of the last token in slice...
+        if result.last().is_some_and(|last| last.end() != range.end()) {
+            // Panic if it's greater than the start offset of the next token. If it's equal to the
+            // start offset, then it's at a token boundary which is valid. If it's less than the
+            // start offset, then it's in the gap between the tokens which is valid as well.
+            if tokens_after_start
+                .get(end)
+                .is_some_and(|next| range.end() > next.start())
+            {
+                panic!("Offset cannot be in between a token")
+            }
+        }
+
+        result
     }
 
     /// Returns a slice of tokens after the given [`TextSize`] offset.
     ///
-    /// If the given offset lies within a token, the returned slice will start from the token after
-    /// that. In other words, the returned slice will not include the token containing the offset.
+    /// If the given offset is between two tokens, the returned slice will start from the following
+    /// token. In other words, if the offset is between the end of previous token and start of next
+    /// token, the returned slice will start from the next token.
+    ///
+    /// # Panics
+    ///
+    /// If the given offset is within a token range.
     pub fn after(&self, offset: TextSize) -> &[Token] {
         let start = self.partition_point(|token| token.start() < offset);
-        &self[start..]
+        let result = &self[start..];
+
+        // If the offset isn't the start of the first token in slice...
+        if result.first().is_some_and(|token| token.start() != offset) {
+            // Panic if it's less than the end offset of the previous token. If it's equal to the
+            // end offset, then it's at a token boundary which is valid. If it's greater than the
+            // end offset, then it's in the gap between the tokens which is valid as well.
+            if self
```

Comment:
I think `saturating_sub` here is incorrect when the file contains a BOM header, in which case the first token starts at offset 1 (or 2?). 

For `offset == 0`:

* `start == 0` 
* `token.start() == 1` (and, therefore, `offset != token.start()`
* `saturating_sub` returns `0` (again the first token)
* `offset` < `token.end()` and panic, where it shouldn't. I think we need to manually test here that `start > 0 && self.get(start - 1).is_some_and...

Reply:
Good catch. I just checked, it starts at offset 3:
```
3 Name {
    name: "hello",
} 8
8 Newline 8
```